Add section to GOI docs with details on mapping genes

cbethell opened the pull request:
**Issue Addressed**
Closes #265

**What is the purpose of these changes?**
<!--Provide some background on the changes proposed.-->
This PR adds a separate section to the `optional-goi-analysis/README.md` file that provides additional details around the optional mapping gene identifiers step of the GOI workflow.

**What changes did you make?**
<!--Describe the concrete changes that you made or additional features that were added, be as detailed as possible in the steps you took.-->
The `optional-goi-analysis/README.md` file is updated here to include a `Gene mapping` section that describes why gene identifier mapping may be necessary and how to adjust the necessary parameters via the command line when running the workflow.

**Any comments, concerns, or questions important for reviewers**
Note that this PR is stacked on the branch in PR #281 
- Are there any additional details here that you believe should be added to this section on mapping genes?

**Checklist**

Place an `x` in all boxes that you have completed.

- [ ] I have run the most recent version of the code
- [ ] If I am adding in reports or generating any plots, I have reviewed the necessary reports
- [x] I have added all necessary documentation (if applicable)

allyhawkins commented:
I realized after I just reviewed #281 that this next PR I was reviewing was the mapping section. I'm going to hold off on reviewing, because I want to flush out the overall organization first and then when that is merged incorporate the mapping a little bit more by framing the parameters as mapping parameters rather than GOI parameters.

cbethell commented:
I've updated this branch with the latest updates to the GOI README and modified the GOI parameters section to be a "Gene mapping parameters" section. Let me know what you think of things now @allyhawkins!

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -122,6 +123,28 @@ The parameters found in the `config/goi_config.yaml` file can be optionally modi
 
+## Gene mapping
+
+If the input gene identifiers do not match the gene identifiers used in the `SingleCellExperiment` object used as input, then input gene identifiers must first be mapped to a specified type.
```
```suggestion
The first step in the workflow is to ensure that the input gene identifiers used in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 
The `SingleCellExperiment` objects that are returned by the core workflow will contain gene names as Ensembl ids. 
```

So here I'm suggesting some ways to rephrase this paragraph to make sure we explicitly state what it means to map and why we need to map. We want to be sure we are clear with the fact that the output from the core workflow contains Ensembl Ids so if you input anything else you need to map.

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -122,6 +123,28 @@ The parameters found in the `config/goi_config.yaml` file can be optionally modi
+## Gene mapping
+
+If the input gene identifiers do not match the gene identifiers used in the `SingleCellExperiment` object used as input, then input gene identifiers must first be mapped to a specified type.
+When the `perform_mapping` flag is set to `TRUE`, users can map the provided gene identifiers to the identifiers used to label the rownames of the `SingleCellExperiment` object.
```
```suggestion
If the provided genes of interest list contains another identifier type (e.g., gene symbol, Entrez id) that does not match the gene names present in the `SingleCellExperiment` objects then mapping must be performed.
The `perform_mapping` flag is used to indicate whether or not gene mapping will be performed and by default is set to `TRUE`.
```

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -122,6 +123,28 @@ The parameters found in the `config/goi_config.yaml` file can be optionally modi
 |[View Genes of Interest Config File](../config/goi_config.yaml)|
 |---|
 
+## Gene mapping
```
I actually think you need to move this up above the section you have on parameters and then have the genes of interest parameters within this section. Having the parameters first is confusing because you have introduced them with little context about what mapping actually is.

allyhawkins reviewed (commented):
Thanks for making the updates to this. My major comment here is about moving the section you have on mapping to be before you introduce the gene mapping parameters. You also want to make sure we have a good explanation of what it really means to map and why you would need it. We should be sure to mention what types of identifiers are present in the output from the core workflow.

cbethell commented on optional-goi-analysis/README.md:
```diff
@@ -103,12 +104,12 @@ snakemake --snakefile goi.snakefile \
 It is **not required** to alter these parameters to run the workflow, but if you would like to modify the gene identifier mapping, you can do so by changing these parameters. 
 
-The parameters found in the `config/goi_config.yaml` file can be optionally modified and are as follows:
+The following gene mapping parameters found in the `config/goi_config.yaml` file can be optionally modified:
```
Good point!

cbethell commented on optional-goi-analysis/README.md:
```diff
@@ -122,6 +123,28 @@ The parameters found in the `config/goi_config.yaml` file can be optionally modi
+  goi_list="<RELATIVE PATH TO YOUR GOI LIST>" \
+  organism="<NAME OF ORGANISM ASSOCIATED WITH THE PROVIDED GOI>" \
+  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN PROVIDED GOI LIST>" \
+  sce_rownames_identifier="<TYPE OF GENE IDENTIFIER IN ROWNAMES OF INPUT SCE OBJECT>" \
+  perform_mapping=TRUE \
```
Another great point! 👍

cbethell commented:
Thanks for the re-review @allyhawkins, those sentence re-structuring suggestions do seem to be much more clear! I've also moved re-arranged the mapping section as you suggested 🚀

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -68,6 +69,44 @@ This `SingleCellExperiment` object must contain a log-normalized counts matrix i
 
+## Gene mapping
+
+The first step in the workflow is to ensure that the input gene identifiers used in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 
```
```suggestion
The first step in the workflow is to ensure that the input gene identifiers provided in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 
```

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -68,6 +69,44 @@ This `SingleCellExperiment` object must contain a log-normalized counts matrix i
+|------------------|-------------|---------------|
+| `organism` | the organism associated with the provided genes and data | "Homo sapiens" |
+| `sce_rownames_identifier` | the type of gene identifiers found in the rownames of the SingleCellExperiment object; note that this parameter should not be changed unless the output of the core analysis workflow has been altered in some way prior to running this module | "ENSEMBL" |
+| `perform_mapping` | a binary value indicating whether or not to perform gene identifier mapping | `TRUE` |
```
Can you move this to be first in the list.

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -68,6 +69,44 @@ This `SingleCellExperiment` object must contain a log-normalized counts matrix i
+### Gene mapping parameters
+
+The [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, sets the defaults for the parameters required for mapping the gene identifiers in the genes of interest workflow.
+It is **not required** to alter these parameters to run the workflow, but if you would like to modify the gene identifier mapping, you can do so by changing these parameters. 
```
```suggestion
These parameters will only be used if `perform_mapping` is set to `TRUE`.
It is **not required** to alter these parameters to run the workflow, but if you would like to modify the gene identifier mapping, you can do so by changing these parameters. 
```

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -68,6 +69,44 @@ This `SingleCellExperiment` object must contain a log-normalized counts matrix i
+| Parameter        | Description | Default value |
+|------------------|-------------|---------------|
+| `organism` | the organism associated with the provided genes and data | "Homo sapiens" |
+| `sce_rownames_identifier` | the type of gene identifiers found in the rownames of the SingleCellExperiment object; note that this parameter should not be changed unless the output of the core analysis workflow has been altered in some way prior to running this module | "ENSEMBL" |
```
```suggestion
| `sce_rownames_identifier` | the type of gene identifiers found in the rownames of the `SingleCellExperiment` object; note that this parameter should not be changed unless the output of the core analysis workflow has been altered in some way prior to running this module | "ENSEMBL" |
```

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -68,6 +69,44 @@ This `SingleCellExperiment` object must contain a log-normalized counts matrix i
+|---|
+
+
 ## Parameters and config file
```
Sorry for the confusion, I meant to just move the gene mapping before the gene mapping parameters not before all the parameters. This should be moved back above to be before `Gene mapping`. 
Also before you make your next commits and move stuff around don't forget to update the TOC.

cbethell replied:
Ah oops, got it!!

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -84,6 +123,7 @@ These include the following parameters:
 | `project_metadata` | relative path to your specific project metadata TSV file (use the same `project_metadata` used in the prerequisite core workflow) |
 | `goi_list` | the file path to a tsv file containing the list of genes that are of interest; the default file path is `example-data/goi-lists/example_goi_list.tsv` |
 | `provided_identifier` | the type of gene identifiers used to populate the genes of interest list; example values that can implemented here include `"ENSEMBL"`, `"ENTREZID"`, `"SYMBOL"` -- where `"SYMBOL"` is the default (see more keytypes [here](https://jorainer.github.io/ensembldb/reference/EnsDb-AnnotationDbi.html)) |
+| `overwrite` | a binary value indicating whether or not to overwrite existing output files | `TRUE` |
```
Here you have the default value for `overwrite` but do not have a separate column for the other rows. Can you add default as a separate column for the entire table.

allyhawkins commented on optional-goi-analysis/README.md:
````diff
@@ -68,6 +69,44 @@ This `SingleCellExperiment` object must contain a log-normalized counts matrix i
+If the provided genes of interest list contains another identifier type (e.g., gene symbol, Entrez id) that does not match the gene names present in the `SingleCellExperiment` objects, then mapping must be performed.
+The `perform_mapping` flag is used to indicate whether or not gene mapping will be performed and by default is set to `TRUE`.
+
+To run the gene mapping step, you can run the below command while using the `--config` flag to tailor the genes of interest mapping parameters mentioned in the [gene mapping parameters section above](#gene-mapping-parameters):
+
+```
+snakemake --snakefile goi.snakefile \ 
+  --cores 2 \
+  --use-conda \
+  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
+  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
+  goi_list="<RELATIVE PATH TO YOUR GENES OF INTEREST LIST>" \
+  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>"
+```
````
I actually think we should change this to show an example of running with mapping and running without mapping.

allyhawkins reviewed (commented):
I think we're getting there! I have a few more structural comments that I hope will make things more clear.

cbethell commented:
Okay I committed some further updates to fix things up -- let me know what you think of the structure and the examples showing how to run the workflow with and without the gene mapping step!

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -99,24 +101,59 @@ snakemake --snakefile goi.snakefile \
+### Gene mapping parameters
 
-We have provided an additional [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml` which sets the defaults for all parameters required for running the genes of interest workflow.
+The [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, sets the defaults for the parameters required for mapping the gene identifiers in the genes of interest workflow.
```
```suggestion
The [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, sets the defaults for additional parameters required for mapping the gene identifiers in the genes of interest workflow.
```

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -99,24 +101,59 @@ snakemake --snakefile goi.snakefile \
+If the provided genes of interest list contains another identifier type (e.g., gene symbol, Entrez id) that does not match the gene names present in the `SingleCellExperiment` objects, then mapping must be performed.
+The `perform_mapping` flag is used to indicate whether or not gene mapping will be performed and by default is set to `TRUE`.
+
+To run the gene mapping step, you can run the below command while using the `--config` flag to tailor the genes of interest mapping parameters mentioned in the [gene mapping parameters section above](#gene-mapping-parameters):
```
```suggestion
To run the workflow with the gene mapping step, you will only need to provide the required parameters mentioned in the [project-specific parameters section](#project-specific-parameters).
```
I think we can simplify this and just have another example here of how to run it with mapping, without modifying anything differently. I was thinking about how we show different examples throughout `scpca-nf` and sometimes we have the same command a few times, and I think that's okay because people will read different sections.

allyhawkins commented on optional-goi-analysis/README.md:
```diff
@@ -99,24 +101,59 @@ snakemake --snakefile goi.snakefile \
+  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
+  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
+  goi_list="<RELATIVE PATH TO YOUR GENES OF INTEREST LIST>" \
+  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>" \
+  perform_mapping=TRUE \
+  organism="<NAME OF ORGANISM ASSOCIATED WITH THE PROVIDED GOI>" \
+  sce_rownames_identifier="<TYPE OF GENE IDENTIFIER IN SCE ROWNAMES>" \
+  multi_mappings="list"
```
```suggestion
  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>"
```
Following my above comment I think we can simplify this.

allyhawkins reviewed (approved):
I'm going to go ahead and approve, given that you accept the last few suggestions.
